### DISCUSSION TITLE
Computation speed for various wrappers

### DISCUSSION THREAD

**Discussion by cooper-david-a** [CoolProp/CoolProp#2358] [General]
Hello everybody,

I'm currently using the javascript wrapper to compute individual states and plot them on a P-H diagram or T-S diagram.  https://daveisthebest.com/ThermoPropertyCalculator/

The diagrams are generated with contour plots so behind the scenes I pre-computed 200 x 200 arrays.

I was hoping to generate these arrays on the fly and expected that c++ would be much faster.  However, after learning the minimum amount of cmake and running a small test, it seems my c++ implementation runs less than half as fast as the javascript in the browser.  

Is this expected?

Before I spend a bunch of time troubleshooting, I was hoping someone could explain this to me or point me in some productive direction.

**Comment by dvd101x** [comment]
Hi, I would like to contribute to the discussion indirectly as I can't run C++

I ran some benchmarks comparing JavaScript and Python for a specific task.

| Wrapper | Type | $\mu \[\mu s\]$ | $\sigma \[\mu s\]$ |
|:---|:---|---:|---:|
| Javascript (wasm) | Local | 376 | 1.03 |
| Jupyter pyodide (wasm) | Local| 162 | 0.24 |
| Jupyter python | Local | 112 | 0.52 |
| Google Colab | Cloud | 180 | 27.1 |


## Testing Methodology

- **Task:** Running `PropsSI('T','P',101325,'Q',0,'Water')`
- **Hardware:** MacBook Air with M1 chip and 8GB RAM (local testing)
- Benchmarking:
  - **Jupyter:** Magic command `%%timeit` resulting in (7 runs, 10,000 loops each)
  - **Javascript:** 7 runs, 10,000 loops each

## Observations

- WebAssembly (wasm) seems to perform well in both JavaScript and Pyodide, offering comparable speed to the native Python wrapper.

**Reply by cooper-david-a** [reply]
Thanks @dvd101x ,
Alright, I'm back on this.  I'm converting the original site over to Angular from vanilla JS.  It's been down for a few weeks but I hope to have it back up soon.  After I get it back up I'll start diving back into wasm.  I'm starting from scratch so it may be slow going.

**Reply by ibell** [reply]
Is the exception handling resolved? That's a non-starter until that is somehow fixed.

**Reply by ibell** [reply]
But that speed looks very promising, better than I had seen before.

**Reply by dvd101x** [reply]
Hi Ian,

They accepted my PR to fix the exception handling. They guided me through all the process.

It hasn't been published yet. I was planning on reporting back upon publishing. 

https://github.com/pyodide/pyodide/pull/4455

**Reply by dvd101x** [reply]
BTW I saw it is working here:
https://pyodide.org/en/latest/console.html

And referenced here:
https://github.com/mgreminger/EngineeringPaper.xyz/commit/db949f48d3b060f74d0c84fc8ed787fed9bad2d6

**Reply by ibell** [reply]
Indeed that's awesome!

**Reply by dvd101x** [reply]
I'm happy to report that the [PR](https://github.com/pyodide/pyodide/pull/4455) was accepted and it's now part of the new release.

https://github.com/pyodide/pyodide/releases/tag/0.26.0

**Reply by cooper-david-a** [reply]
I'd like to finish my little thermo-property calculator just to wrap it up but then I think I'd like to help out on your engineering calculator.

**Comment by ibell** [comment]
This is really great! I think we should consider to sunset the CoolProp
online and give some pointers to people on how to use the pyiodide things

**Reply by dvd101x** [reply]
I'm not familiar with CoolProp online. But maybe some examples with pyodide could be included on the website.

Just noticed that now pyscript includes the fix so now it's easier to use its features.
https://github.com/pyscript/pyscript/releases/tag/2024.6.1

**Reply by mgreminger** [reply]
Now that CoolProp is working properly in Pyodide, I've been able to release an update to EngineeringPaper.xyz that includes CoolProp capabilities. See the demo as an [EngineeringPaper.xyz example sheet](https://engineeringpaper.xyz/h2x75xbLYjEgg7uWgXcHvz) or as a [YouTube video](https://youtu.be/6QL-jvBdoSg?si=tkZI2Wfiz08KwfUr). Many thanks to @dvd101x for working on getting CoolProp working in Pyodide and for letting me know about CoolProp. Many thanks to the CoolProp team as well for making this excellent open source library!

**Reply by cooper-david-a** [reply]
Engineeringpaper.xyz is impressive.  When I set out on this journey, my intent was to eventually develop something like it.  Now that I know it exists, I think it would be a better use of my time to help out in some way.
